src/PhpSpreadsheet/Worksheet/Table.php

```diff
@@ -0,0 +1,471 @@
+<?php
+
+namespace PhpOffice\PhpSpreadsheet\Worksheet;
+
+use PhpOffice\PhpSpreadsheet\Cell\Coordinate;
+use PhpOffice\PhpSpreadsheet\Exception as PhpSpreadsheetException;
+use PhpOffice\PhpSpreadsheet\Worksheet\Table\TableStyle;
+
+class Table
+{
+    /**
+     * Table Name.
+     *
+     * @var string
+     */
+    private $name = '';
+
+    /**
+     * Show Header Row.
+     *
+     * @var bool
+     */
+    private $showHeaderRow = true;
+
+    /**
+     * Show Totals Row.
+     *
+     * @var bool
+     */
+    private $showTotalsRow = false;
+
+    /**
+     * Table Range.
+     *
+     * @var string
+     */
+    private $range = '';
+
+    /**
+     * Table Worksheet.
+     *
+     * @var null|Worksheet
+     */
+    private $workSheet;
+
+    /**
+     * Table Column.
+     *
+     * @var Table\Column[]
+     */
+    private $columns = [];
+
+    /**
+     * Table Style.
+     *
+     * @var TableStyle
+     */
+    private $style;
+
+    /**
+     * Create a new Table.
+     *
+     * @param string $range (e.g. A1:D4)
+     *
+     * @return $this
+     */
+    public function __construct(string $range = '', ?Worksheet $worksheet = null)
+    {
+        $this->setRange($range);
+        $this->setWorksheet($worksheet);
+        $this->style = new TableStyle();
+    }
+
+    /**
+     * Get Table name.
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /**
+     * Set Table name.
+     *
+     * @return $this
+     */
+    public function setName(string $name)
+    {
+        $name = trim($name);
+
+        if (strlen($name) == 1 && in_array($name, ['C', 'c', 'R', 'r'])) {
+            throw new PhpSpreadsheetException('The table name is invalid');
+        }
+        if (strlen($name) > 255) {
+            throw new PhpSpreadsheetException('The table name cannot be longer than 255 characters');
+        }
+        // Check for A1 or R1C1 cell reference notation
+        if (
+            preg_match(Coordinate::A1_COORDINATE_REGEX, $name) ||
+            preg_match('/^R\[?\-?[0-9]*\]?C\[?\-?[0-9]*\]?$/i', $name)
+        ) {
+            throw new PhpSpreadsheetException('The table name can\'t be the same as a cell reference');
+        }
+        if (!preg_match('/^[A-Z_\\\\]/i', $name)) {
+            throw new PhpSpreadsheetException('The table name must begin a name with a letter, an underscore character (_), or a backslash (\)');
+        }
+        if (!preg_match('/^[A-Z_\\\\][A-Z0-9\._]+$/i', $name)) {
+            throw new PhpSpreadsheetException('The table name contains invalid characters');
+        }
+
+        $this->name = $name;
+
+        return $this;
+    }
+
+    /**
+     * Get show Header Row.
+     *
+     * @return bool
+     */
+    public function getShowHeaderRow()
```

Comment:
Where possible, we're trying to improve the codebase by using return typehints rather than an `@return` entr in the docblock; though this has only been possible since PHP started supporting return typehints, so a lot of the older code still hasn't been updated

Reply:
Changed wherever possible  44d63f027aedc865342e1f87e30234c91a5df359